pkg/icingadb/sla_lifecycle.go

```diff
@@ -0,0 +1,95 @@
+package icingadb
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icingadb/pkg/contracts"
+	"github.com/icinga/icingadb/pkg/icingadb/objectpacker"
+	v1 "github.com/icinga/icingadb/pkg/icingadb/v1"
+	"github.com/icinga/icingadb/pkg/types"
+	"github.com/icinga/icingadb/pkg/utils"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+	"time"
+)
+
+// GetCheckableFromSlaLifecycle returns the original checkable from which the specified sla lifecycle were transformed.
+// When the passed entity is not of type *SlaLifecycle, it is returned as is.
+func GetCheckableFromSlaLifecycle(e contracts.Entity) contracts.Entity {
+	s, ok := e.(*v1.SlaLifecycle)
+	if !ok {
+		return e
+	}
+
+	return s.SourceEntity
+}
+
+// CreateSlaLifecyclesFromCheckables transforms the given checkables to sla lifecycle struct
+// and streams them into a returned channel.
+func CreateSlaLifecyclesFromCheckables(
+	ctx context.Context, g *errgroup.Group, db *DB, subject contracts.Entity, entities <-chan contracts.Entity, isDeleteEvent bool,
+) <-chan contracts.Entity {
+	slaLifecycles := make(chan contracts.Entity, 1)
+
+	g.Go(func() error {
+		defer close(slaLifecycles)
+
+		env, ok := v1.EnvironmentFromContext(ctx)
+		if !ok {
+			return errors.New("can't get environment from context")
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case checkable, ok := <-entities:
+				if !ok {
+					return nil
+				}
+
+				now := time.Now()
+
+				sl := &v1.SlaLifecycle{
+					EnvironmentMeta: v1.EnvironmentMeta{EnvironmentId: env.Id},
+					CreateTime:      types.UnixMilli(now),
+					DeleteTime:      types.UnixMilli(time.Unix(0, 0)),
+					SourceEntity:    checkable,
+				}
+
+				switch subject.(type) {
+				case *v1.Host:
+					sl.HostId = checkable.(v1.BinaryIDer).BinaryID()
+					sl.Id = utils.Checksum(objectpacker.MustPackSlice(sl.EnvironmentId, sl.HostId))
+				case *v1.Service:
+					sl.HostId = checkable.(v1.BinaryHostIDer).BinaryHostID()
+					sl.ServiceId = checkable.(v1.BinaryIDer).BinaryID()
+					if !sl.HostId.Valid() {
+						err := db.QueryRowxContext(
+							ctx, db.Rebind(fmt.Sprintf("%s %s", db.BuildSelectStmt(checkable, v1.HostMeta{}), ` WHERE "id" = ?`)),
+							sl.ServiceId,
+						).StructScan(sl)
+
+						if err != nil {
+							return err
+						}
+					}
```

Comment:
For delete event of runtime updates.

Reply:
I think it might make sense doing this in the runtime update specific code if it's only for one very specific situation.

This should (hopefully, I haven't checked if anything else might prevent this) also allow to use the existing chunks/bulks that are generated there and query the information for all entities in there in a single query.

Reply:
> I think it might make sense doing this in the runtime update specific code if it's only for one very specific situation.

https://github.com/Icinga/icingadb/pull/566#discussion_r1140133758

> This should (hopefully, I haven't checked if anything else might prevent this) also allow to use the existing chunks/bulks that are generated there and query the information for all entities in there in a single query.

How is this supposed to work? The chunks are generated within the respective `DB#*BulkExec()` methods, which you don't have access to. Btw, this is just relevant for delete events of the runtime updates, which are IMAO not that expensive as might you think.

Reply:
Ok, probably ignore that comment for now. Looks like I should look into this part in more detail first.

Reply:
> > I think it might make sense doing this in the runtime update specific code if it's only for one very specific situation.
> 
> [#566 (comment)](https://github.com/Icinga/icingadb/pull/566#discussion_r1140133758)

That select is just untypical for Icinga DB, so it will probably be a bit weird everywhere. Probably indeed makes more sense here, but there should be a clear comment what this is for.

Also, is there an option for Icinga 2 to send this information directly? I don't mean as an replacement for this bit of code but as an optimization that would make this just a fallback?
 
> > This should (hopefully, I haven't checked if anything else might prevent this) also allow to use the existing chunks/bulks that are generated there and query the information for all entities in there in a single query.
> 
> How is this supposed to work? The chunks are generated within the respective `DB#*BulkExec()` methods, which you don't have access to.

My idea was that the `OnSuccess` functions all receive a slice and the `host_id` could have been queried for all in there as well.

> Btw, this is just relevant for delete events of the runtime updates, which are IMAO not that expensive as might you think.

This only affects runtime deletion of services, not something that's super common to do very often (things would be different here for other types like downtimes).

Reply:
> My idea was that the `OnSuccess` functions all receive a slice and the `host_id` could have been queried for all in there as well.

The `OnSuccess` functions do all get a slice, but since the `SlaLifecycle` entries are processed first, this won't be useful in any way, as the `host_id` has to be set beforehand.

> Also, is there an option for Icinga 2 to send this information directly?

There is of course an option to set this information directly by Icinga 2, but none of you guys wanted a change on Icinga 2 side in this regard.

Reply:
> but none of you guys wanted a change on Icinga 2 side in this regard.

Not as an requirement. But allowing for an optimization would be good. So my question is: with the current state of the PR, if there was a `host_id` in the Redis runtime update stream, would this be used and skip the query? Just so that we have the option to do this as an optimization later, without requiring it immediately.

Reply:
> if there was a `host_id` in the Redis runtime update stream, would this be used and skip the query?

Yes, this would work if the Redis runtime **delete** stream contained `host_id`.